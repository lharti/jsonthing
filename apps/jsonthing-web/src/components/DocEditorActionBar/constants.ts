export enum SaveStatus {
    IDLE = 'idle',
    PENDING = 'pending',
    SAVED = 'saved',

    NOT_ALLOWED = 'not_allowed',
}

export const SAVE_BTN_LABELS: Record<SaveStatus, string> = {
    [SaveStatus.IDLE]: 'Save',
    [SaveStatus.PENDING]: 'Saving...',
    [SaveStatus.SAVED]: 'Saved',

    [SaveStatus.NOT_ALLOWED]: 'Cannot save',
}
