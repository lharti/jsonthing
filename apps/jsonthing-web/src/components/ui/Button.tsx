import { cn } from '@/lib/utils'
import { Slot } from '@radix-ui/react-slot'
import { cva, type VariantProps } from 'class-variance-authority'
import * as React from 'react'

const buttonVariants = cva(
    `
      inline-flex items-center justify-center whitespace-nowrap rounded-md
      text-sm font-medium transition-colors

      dark:focus-visible:ring-gray-300

      disabled:pointer-events-none disabled:opacity-50

      focus-visible:outline-none focus-visible:ring-1
      focus-visible:ring-gray-950
    `,
    {
        variants: {
            variant: {
                default: `
                  bg-gray-900 text-gray-50 shadow

                  dark:bg-gray-50 dark:text-gray-900 dark:hover:bg-gray-50/90

                  hover:bg-gray-900/90
                `,
                destructive: `
                  bg-red-500 text-gray-50 shadow-sm

                  dark:bg-red-900 dark:text-gray-50 dark:hover:bg-red-900/90

                  hover:bg-red-500/90
                `,
                outline: `
                  border border-gray-200 bg-white shadow-sm

                  dark:border-gray-800 dark:bg-gray-950 dark:hover:bg-gray-800
                  dark:hover:text-gray-50

                  hover:bg-gray-100 hover:text-gray-900
                `,
                secondary: `
                  bg-gray-100 text-gray-900 shadow-sm

                  dark:bg-gray-800 dark:text-gray-50 dark:hover:bg-gray-800/80

                  hover:bg-gray-100/80
                `,
                ghost: `
                  dark:hover:bg-gray-800 dark:hover:text-gray-50

                  hover:bg-gray-100 hover:text-gray-900
                `,
                link: `
                  text-gray-900 underline-offset-4

                  dark:text-gray-50

                  hover:underline
                `,
            },
            size: {
                default: 'h-9 px-4 py-2',
                sm: 'h-8 rounded-md px-3 text-xs',
                lg: 'h-10 rounded-md px-8',
                icon: 'size-9',
            },
        },
        defaultVariants: {
            variant: 'default',
            size: 'default',
        },
    },
)

export interface ButtonProps
    extends React.ButtonHTMLAttributes<HTMLButtonElement>,
        VariantProps<typeof buttonVariants> {
    readonly asChild?: boolean
}

const Button = React.forwardRef<HTMLButtonElement, ButtonProps>(
    ({ className, variant, size, asChild = false, ...props }, ref) => {
        const Comp = asChild ? Slot : 'button'
        return (
            <Comp
                className={cn(buttonVariants({ variant, size, className }))}
                ref={ref}
                {...props}
            />
        )
    },
)

Button.displayName = 'Button'

export { Button, buttonVariants }
