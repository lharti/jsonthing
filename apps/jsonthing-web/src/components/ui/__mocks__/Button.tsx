const Button = jest.fn()

export { Button }
