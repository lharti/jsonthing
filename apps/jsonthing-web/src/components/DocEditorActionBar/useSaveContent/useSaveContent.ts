import { useUpdateDoc } from '@/hooks/useUpdateDoc'
import { useEffect, useRef, useState } from 'react'
import { SaveStatus } from '../constants'

interface SavePayload {
    content: string
    id: string
}

interface Options {
    onSuccess?: () => void
    onError?: () => void
}

type UseSaveContent = (
    savePayload: SavePayload,
    options?: Options,
) => {
    save: () => void
    status: SaveStatus
}

export const useSaveContent: UseSaveContent = (
    { content, id },

    options,
) => {
    const [status, setStatus] = useState(SaveStatus.SAVED)

    const savedContent = useRef(content)

    const { updateDoc } = useUpdateDoc()

    const save = () => {
        setStatus(SaveStatus.PENDING)

        updateDoc(
            {
                id,

                payload: {
                    content,
                },
            },

            {
                onSuccess: () => {
                    setStatus(SaveStatus.SAVED)

                    savedContent.current = content

                    options?.onSuccess?.()
                },

                onError: () => {
                    setStatus(SaveStatus.IDLE)

                    options?.onError?.()
                },
            },
        )
    }

    useEffect(() => {
        if (savedContent.current === content) {
            setStatus(SaveStatus.SAVED)
        } else {
            setStatus(SaveStatus.IDLE)
        }
    }, [content])

    return {
        save,
        status,
    }
}
