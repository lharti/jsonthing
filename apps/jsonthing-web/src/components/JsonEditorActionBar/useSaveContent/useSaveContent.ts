import { useUpdateDoc } from '@/hooks/useUpdateDoc'
import { useEffect, useRef, useState } from 'react'
import { SaveStatus } from '../constants'

interface SavePayload {
    content: string
    id: string
}

interface Options {
    onSuccess?: () => void
    onError?: () => void
}

type UseSaveContent = (
    savePayload: SavePayload,
    options?: Options,
) => {
    save: () => void
    status: SaveStatus
}

export const useSaveContent: UseSaveContent = (
    { content, id },

    options,
) => {
    const [status, setStatus] = useState(SaveStatus.SAVED)

    const { updateDoc } = useUpdateDoc()

    const save = () => {
        setStatus(SaveStatus.PENDING)

        updateDoc(
            {
                id,

                payload: {
                    content,
                },
            },

            {
                onSuccess: () => {
                    setStatus(SaveStatus.SAVED)

                    options?.onSuccess?.()
                },

                onError: () => {
                    setStatus(SaveStatus.IDLE)

                    options?.onError?.()
                },
            },
        )
    }

    const isFirstRender = useRef(true)

    useEffect(() => {
        if (isFirstRender.current) {
            isFirstRender.current = false

            return
        }

        setStatus(SaveStatus.IDLE)
    }, [content])

    return {
        save,
        status,
    }
}
