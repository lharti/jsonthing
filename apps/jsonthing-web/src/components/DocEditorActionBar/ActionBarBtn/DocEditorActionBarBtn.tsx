import { Button, ButtonProps } from '@/components/ui/Button'
import {
    Tooltip,
    TooltipContent,
    TooltipTrigger,
} from '@/components/ui/Tooltip'
import { IconLoader, IconProps } from '@tabler/icons-react'
import React from 'react'

export interface DocEditorActionBarBtnProps extends ButtonProps {
    label: string

    Icon: React.ComponentType<IconProps>

    onClick: () => void

    iconOnly?: boolean

    isLoading?: boolean

    isSuccess?: boolean

    isError?: boolean
}

export const DocEditorActionBarBtn: React.FC<DocEditorActionBarBtnProps> = ({
    label,
    Icon,
    variant,

    onClick,

    iconOnly,
    isLoading,

    ...buttonProps
}) => {
    const ButtonIcon = () =>
        isLoading ? (
            <IconLoader className="mr-1 animate-spin" />
        ) : (
            <Icon className="mr-1" />
        )

    return (
        <Tooltip>
            <TooltipTrigger asChild>
                <Button
                    variant={variant || 'ghost'}
                    size="sm"
                    role="button"
                    aria-label={label}
                    onClick={onClick}
                    {...buttonProps}
                >
                    <ButtonIcon />

                    {iconOnly ? '' : label}
                </Button>
            </TooltipTrigger>

            <TooltipContent>{label}</TooltipContent>
        </Tooltip>
    )
}
