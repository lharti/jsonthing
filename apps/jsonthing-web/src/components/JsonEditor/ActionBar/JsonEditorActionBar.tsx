import { Button } from '@/components/ui/Button'
import { prettifyJson } from '@/lib/prettify-json'
import React from 'react'

export interface JsonEditorActionBarProps {
    className?: string

    editorContent: string
    setEditorContent: (value: string) => void
}

export const JsonEditorActionBar: React.FC<JsonEditorActionBarProps> = ({
    className,

    editorContent,
    setEditorContent,
}) => {
    return (
        <div className={className}>
            <Button
                variant="outline"
                className="my-2 h-6 rounded-none bg-transparent p-0 px-2"
                onClick={() => {
                    const prettyEditorContent = prettifyJson(editorContent)

                    setEditorContent(prettyEditorContent)
                }}
            >
                {'Prettify'}
            </Button>
        </div>
    )
}
