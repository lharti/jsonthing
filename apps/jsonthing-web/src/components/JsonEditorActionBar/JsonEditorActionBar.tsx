import { useUpdateDoc } from '@/hooks/useUpdateDoc'
import { prettifyJson } from '@/lib/prettify-json'
import { cn } from '@/lib/utils'
import {
    IconClipboardCopy,
    IconDeviceFloppy,
    IconWand,
} from '@tabler/icons-react'
import { useParams } from 'next/navigation'
import React from 'react'
import { JsonEditorActionBarBtn } from './ActionBarBtn'

export interface JsonEditorActionBarProps {
    className?: string

    editorContent: string
    setEditorContent: React.Dispatch<React.SetStateAction<string>>
}

export const JsonEditorActionBar: React.FC<JsonEditorActionBarProps> = ({
    className,

    editorContent,
    setEditorContent,
}) => {
    const prettifyContent = () => {
        setEditorContent(prettifyJson(editorContent))
    }

    const copyToClipboard = () => {
        navigator.clipboard.writeText(editorContent)
    }

    const { id } = useParams()

    const { updateDoc } = useUpdateDoc()

    const save = () => {
        updateDoc({
            id: id as string,

            payload: {
                content: editorContent,
            },
        })
    }

    return (
        <div className={cn('flex', className)}>
            <JsonEditorActionBarBtn
                iconOnly
                label="Prettify"
                Icon={IconWand}
                onClick={() => prettifyContent()}
            />

            <JsonEditorActionBarBtn
                iconOnly
                label="Copy"
                Icon={IconClipboardCopy}
                onClick={() => copyToClipboard()}
            />
            <JsonEditorActionBarBtn
                variant="outline"
                label="Save"
                Icon={IconDeviceFloppy}
                onClick={() => save()}
            />
        </div>
    )
}
