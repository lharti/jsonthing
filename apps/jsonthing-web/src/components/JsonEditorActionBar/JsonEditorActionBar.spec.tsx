import { Button } from '@/components/ui/Button'
import {
    IconClipboardCopy,
    IconDeviceFloppy,
    IconWand,
} from '@tabler/icons-react'
import { render, screen } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import React from 'react'
import { JsonEditorActionBarBtn } from './ActionBarBtn'
import { SaveStatus } from './constants'
import { JsonEditorActionBar } from './JsonEditorActionBar'
import { useSaveContent } from './useSaveContent'

jest.mock('./ActionBarBtn')
const JsonEditorActionBarBtnMock = jest.mocked(JsonEditorActionBarBtn)

const mockActionBarBtn = () => {
    return JsonEditorActionBarBtnMock.mockImplementation(
        ({ onClick, label }) => (
            <Button role="button" aria-label={label} onClick={onClick} />
        ),
    )
}

// eslint-disable-next-line jest/no-untyped-mock-factory
jest.mock('next/navigation', () => ({
    useParams: jest.fn(() => ({ id: 'DOC_ID' })),
}))

jest.mock('./useSaveContent')

const useSaveContentMock = jest.mocked(useSaveContent).mockReturnValue({
    save: jest.fn(),
    status: SaveStatus.SAVED,
})

describe('<JsonEditorActionBar />', () => {
    it('should render', () => {
        expect.assertions(1)

        mockActionBarBtn()

        const { container } = render(
            <JsonEditorActionBar
                editorContent="{}"
                setEditorContent={jest.fn()}
            />,
        )

        expect(container).toMatchSnapshot('<JsonEditorActionBar />')
    })

    it('should setup useSaveContent', () => {
        expect.assertions(1)

        render(
            <JsonEditorActionBar
                editorContent="{}"
                setEditorContent={jest.fn()}
            />,
        )

        expect(useSaveContentMock).toHaveBeenCalledWith(
            {
                id: 'DOC_ID',
                content: '{}',
            },

            {
                onError: expect.any(Function),
            },
        )
    })

    describe('prettify button', () => {
        it('should be setup correctly', () => {
            expect.assertions(1)

            const ActionBtnMock = mockActionBarBtn()

            render(
                <JsonEditorActionBar
                    editorContent="{}"
                    setEditorContent={jest.fn()}
                />,
            )

            expect(ActionBtnMock).toHaveBeenCalledWith(
                {
                    iconOnly: true,
                    label: 'Prettify',
                    Icon: IconWand,
                    onClick: expect.any(Function),
                },

                undefined,
            )
        })

        it('should prettify editor content when clicked', () => {
            expect.assertions(1)

            mockActionBarBtn()

            const uglyJson = `
            {"a":1,"b":2,
            "c": 3}`

            const setEditorContent = jest.fn(value => value)

            render(
                <JsonEditorActionBar
                    editorContent={uglyJson}
                    setEditorContent={setEditorContent}
                />,
            )

            const prettifyBtn = screen.getByRole('button', {
                name: 'Prettify',
            })

            prettifyBtn.click()

            const newEditorContent = setEditorContent.mock.results[0].value

            expect(newEditorContent).toMatchInlineSnapshot(`
                            "{
                              "a": 1,
                              "b": 2,
                              "c": 3
                            }"
                    `)
        })
    })

    describe('copy button', () => {
        it('should be setup correctly', () => {
            expect.assertions(1)

            const ActionBtnMock = mockActionBarBtn()

            render(
                <JsonEditorActionBar
                    editorContent="{}"
                    setEditorContent={jest.fn()}
                />,
            )

            expect(ActionBtnMock).toHaveBeenCalledWith(
                {
                    iconOnly: true,
                    label: 'Copy',
                    Icon: IconClipboardCopy,
                    onClick: expect.any(Function),
                },

                undefined,
            )
        })

        it('should copy editor content to clipboard', async () => {
            expect.assertions(1)

            userEvent.setup()

            mockActionBarBtn()

            const editorContent = `{"random": ${Math.random()}`

            render(
                <JsonEditorActionBar
                    editorContent={editorContent}
                    setEditorContent={jest.fn()}
                />,
            )

            const copyBtn = screen.getByRole('button', {
                name: 'Copy',
            })

            copyBtn.click()

            const clipboardText = await navigator.clipboard.readText()

            expect(clipboardText).toStrictEqual(editorContent)
        })
    })

    describe('save button', () => {
        it('should be setup correctly', () => {
            expect.assertions(1)

            const ActionBtnMock = mockActionBarBtn()

            render(
                <JsonEditorActionBar
                    editorContent="{}"
                    setEditorContent={jest.fn()}
                />,
            )

            expect(ActionBtnMock).toHaveBeenCalledWith(
                {
                    variant: 'outline',
                    label: 'Saved',
                    Icon: IconDeviceFloppy,

                    onClick: expect.any(Function),

                    isLoading: false,
                    disabled: true,
                },

                undefined,
            )
        })

        it('should save content onClick', () => {
            expect.assertions(1)

            const saveMock = jest.fn()

            useSaveContentMock.mockReturnValue({
                save: saveMock,
                status: SaveStatus.IDLE,
            })

            const editorContent = Math.random().toString()

            render(
                <JsonEditorActionBar
                    editorContent={editorContent}
                    setEditorContent={jest.fn()}
                />,
            )

            const saveBtn = screen.getByRole('button', {
                name: 'Save',
            })

            saveBtn.click()

            expect(saveMock).toHaveBeenCalledOnce()
        })
    })
})
