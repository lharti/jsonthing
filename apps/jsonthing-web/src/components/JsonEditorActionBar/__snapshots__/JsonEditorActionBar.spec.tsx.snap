// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<JsonEditorActionBar /> should render: <JsonEditorActionBar /> 1`] = `
<div>
  <div
    class="flex"
  >
    <button
      aria-label="Prettify"
      class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium transition-colors dark:focus-visible:ring-gray-300 disabled:pointer-events-none disabled:opacity-50 focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-gray-950 bg-gray-900 text-gray-50 shadow dark:bg-gray-50 dark:text-gray-900 dark:hover:bg-gray-50/90 hover:bg-gray-900/90 h-9 px-4 py-2"
      role="button"
    />
    <button
      aria-label="Copy"
      class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium transition-colors dark:focus-visible:ring-gray-300 disabled:pointer-events-none disabled:opacity-50 focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-gray-950 bg-gray-900 text-gray-50 shadow dark:bg-gray-50 dark:text-gray-900 dark:hover:bg-gray-50/90 hover:bg-gray-900/90 h-9 px-4 py-2"
      role="button"
    />
    <button
      aria-label="Saved"
      class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium transition-colors dark:focus-visible:ring-gray-300 disabled:pointer-events-none disabled:opacity-50 focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-gray-950 bg-gray-900 text-gray-50 shadow dark:bg-gray-50 dark:text-gray-900 dark:hover:bg-gray-50/90 hover:bg-gray-900/90 h-9 px-4 py-2"
      role="button"
    />
  </div>
</div>
`;
